File: task/linkedlist/MergeTwoSortedLists.java
package linkedlist;

import fastandslowpointers.addTwoNumbers.ListNode;

public class MergeTwoSortedLists {

    public static void main(String[] args) {
        ListNode headA1 = new ListNode(1, new ListNode(2, new ListNode(4)));
        ListNode headB1 = new ListNode(1, new ListNode(3, new ListNode(4)));
        ListNode intersectionNode1 = mergeTwoLists(headA1, headB1);
        System.out.println(intersectionNode1);

        ListNode headA = new ListNode(2, new ListNode(4, new ListNode(6, new ListNode(7, new ListNode(8)))));
        ListNode headB = new ListNode(1, new ListNode(2, new ListNode(3)));
        ListNode intersectionNode = mergeTwoLists(headA, headB);
        System.out.println(intersectionNode);
    }

    public static ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null && l2 == null) return null;
        if (l1 == null) return l2;
        if (l2 == null) return l1;

        var result = new ListNode(0);
        var dummy = result;

        while (l1 != null || l2 != null) {
            if (l1 != null && (l2 == null || l1.val < l2.val)) {
                dummy.next = l1;
                l1 = l1.next;
            } else {
                dummy.next = l2;
                l2 = l2.next;
            }
            dummy = dummy.next;
        }

        return result.next;
    }

    public static ListNode mergeTwoListsRec(ListNode l1, ListNode l2) {
        if (l1 == null && l2 == null) return null;
        if (l1 == null) return l2;
        if (l2 == null) return l1;

        if (l1.val < l2.val) {
            l1.next = mergeTwoListsRec(l1.next, l2);
            return l1;
        } else {
            l2.next = mergeTwoListsRec(l1, l2.next);
            return l2;
        }
    }
}

File: python/week1/MergeTwoSortedLists.py
from typing import Optional

from ListNode import ListNode


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1

        dummy = ListNode()
        result = dummy
        while list1 and list2:
            if list1.val < list2.val:
                dummy.next = list1
                list1 = list1.next
            else:
                dummy.next = list2
                list2 = list2.next
            dummy = dummy.next

        if list1:
            result.next = list1
        else:
            result.next = list2

        return result


if __name__ == '__main__':
    Solution().mergeTwoLists(ListNode(1),
                             ListNode(2))
    Solution().mergeTwoLists(ListNode(1, ListNode(2, ListNode(4))),
                             ListNode(1, ListNode(3, ListNode(4))))
